components/omega/src/ocn/OceanState.cpp

```diff
@@ -0,0 +1,385 @@
+
+#include "OceanState.h"
+#include "DataTypes.h"
+#include "Decomp.h"
+#include "Halo.h"
+#include "IO.h"
+#include "IOField.h"
+#include "Logging.h"
+#include "MachEnv.h"
+#include "MetaData.h"
+#include "OmegaKokkos.h"
+
+namespace OMEGA {
+
+// create the static class members
+OceanState *OceanState::DefaultOceanState = nullptr;
+std::map<std::string, OceanState> OceanState::AllOceanStates;
+
+//------------------------------------------------------------------------------
+// Initialize the state. Assumes that Decomp has already been initialized.
+
+int OceanState::init() {
+
+   int Err = 0; // default successful return code
+
+   // Retrieve the default decomposition and mesh
+   Decomp *DefDecomp     = Decomp::getDefault();
+   HorzMesh *DefHorzMesh = HorzMesh::getDefault();
+   Halo *DefHalo         = Halo::getDefault();
+
+   // These hard-wired variables need to be updated
+   // with retrivals/config options
+   int NTimeLevels = 2;
+   int NVertLevels = 60;
+
+   // Create the default state
+   OceanState DefOceanState("Default", DefHorzMesh, DefDecomp, DefHalo,
+                            NVertLevels, NTimeLevels);
+
+   // Retrieve this mesh and set pointer to DefaultOceanState
+   OceanState::DefaultOceanState = OceanState::get("Default");
+   return Err;
+}
+
+//------------------------------------------------------------------------------
+// Construct a new local state given a decomposition
+
+OceanState::OceanState(
+    const std::string &Name, //< [in] Name for new state
+    HorzMesh *Mesh,          //< [in] HorzMesh for state
+    Decomp *MeshDecomp,      //< [in] Decomp for Mesh
+    Halo *MeshHalo_,         //< [in] Halo for Mesh
+    const int NVertLevels_,  //< [in] number of vertical levels
+    const int NTimeLevels_   //< [in] number of time levels
+) {
+
+   // Retrieve mesh cell/edge/vertex totals from Decomp
+   NCellsOwned = Mesh->NCellsOwned;
+   NCellsAll   = Mesh->NCellsAll;
+   NCellsSize  = Mesh->NCellsSize;
+
+   NEdgesOwned = Mesh->NEdgesOwned;
+   NEdgesAll   = Mesh->NEdgesAll;
+   NEdgesSize  = Mesh->NEdgesSize;
+
+   NVertLevels = NVertLevels_;
+   NTimeLevels = NTimeLevels_;
+
+   MeshHalo = MeshHalo_;
+
+   StateFileName = Mesh->MeshFileName;
+
+   // Allocate state host arrays
+   for (int I = 0; I < NTimeLevels; I++) {
+      LayerThicknessH[I] = HostArray2DR8("LayerThickness" + std::to_string(I),
+                                         NCellsSize, NVertLevels);
+      NormalVelocityH[I] = HostArray2DR8("NormalVelocity" + std::to_string(I),
+                                         NEdgesSize, NVertLevels);
+   }
+
+   // Create state device arrays
+   for (int I = 0; I < NTimeLevels; I++) {
+      LayerThickness[I] = Array2DR8("LayerThickness" + std::to_string(I),
+                                    NCellsSize, NVertLevels);
+      NormalVelocity[I] = Array2DR8("NormalVelocity" + std::to_string(I),
+                                    NEdgesSize, NVertLevels);
+   }
```

Comment:
I know that the host arrays are meant to be temporary, but doing  
```suggestion
   // Create state device arrays
   for (int I = 0; I < NTimeLevels; I++) {
      LayerThickness[I] = createDeviceMirrorCopy(LayerThicknessH[I]);
      NormalVelocity[I] = createDeviceMirrorCopy(NormalVelocityH[I]); 
   }
```
moving this block of code after the read, and removing the explicit copy would prevent creating two separate allocations on CPU.

Reply:
Thanks for the suggestion. I've updated this.